google-apis-core/lib/google/apis/core/storage_upload.rb

```diff
@@ -0,0 +1,184 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'google/apis/core/http_command'
+require 'google/apis/core/api_command'
+require 'google/apis/errors'
+require 'tempfile'
+require 'mini_mime'
+
+module Google
+  module Apis
+    module Core
+      # Base upload command. Not intended to be used directly
+      # @private
+      class StorageUploadCommand < ApiCommand
+        CONTENT_LENGTH_HEADER = "Content-Length"
+        CONTENT_TYPE_HEADER = "Content-Type"
+        UPLOAD_CONTENT_TYPE_HEADER = "X-Upload-Content-Type"
+        LOCATION_HEADER = "Location"
+        CONTENT_RANGE_HEADER = "Content-Range"
+        RESUMABLE = "resumable"
+        OK_STATUS = 200
+        CHUNK_SIZE = 8 * 1024 * 1024 # 8 MB
+
+        # File name or IO containing the content to upload
+        # @return [String, File, #read]
+        attr_accessor :upload_source
+
+        # Content type of the upload material
+        # @return [String]
+        attr_accessor :upload_content_type
+
+        # Content, as UploadIO
+        # @return [Google::Apis::Core::UploadIO]
+        attr_accessor :upload_io
+
+        # Ensure the content is readable and wrapped in an IO instance.
+        #
+        # @return [void]
+        # @raise [Google::Apis::ClientError] if upload source is invalid
+        def prepare!
+          @upload_url = nil
+          @offset = 0
+          @upload_incomplete = true
+          # Prevent the command from populating the body with form encoding, by
+          # asserting that it already has a body. Form encoding is never used
+          # by upload requests.
+          self.body = '' unless self.body
+
+          super
+          if streamable?(upload_source)
+            self.upload_io = upload_source
+            @close_io_on_finish = false
+          elsif self.upload_source.is_a?(String)
+            self.upload_io = File.new(upload_source, 'r')
+            if self.upload_content_type.nil?
+              type = MiniMime.lookup_by_filename(upload_source)
+              self.upload_content_type = type&.content_type
+            end
+            @close_io_on_finish = true
+          else
+            fail Google::Apis::ClientError, 'Invalid upload source'
+          end
+          if self.upload_content_type.nil? || self.upload_content_type.empty?
+            self.upload_content_type = 'application/octet-stream'
```

Comment:
GCS handles the default value for users; `X-Goog-Upload-Content-Type` is optional unless that's changed recently.